Chapter12/Exercise08.Tests/MainWindowTests.cs

```diff
@@ -0,0 +1,189 @@
+        public void _1_ShouldHaveTwoTextBlocksForAreaAndError()
+        {
+            AssertHaveTwoTextBlocksProperlyInitialized();
+        }
+
+        [MonitoredTest("Should have a Max function with 3 parameters"), Order(2)]
+        public void _2_ShouldHaveAMaxFunction()
+        {
+            AssertHasMaxFunction();
+        }
+
+        [MonitoredTest("Max function should compute the greatest of 3 int parameters"), Order(3)]
+        public void _3_MaxShouldWorkProperly()
+        {
+            AssertHasMaxFunction();
+
+            AssertMaxInvoke(10, 10, 2, 3);
+            AssertMaxInvoke(100, 10, 100, 3);
+            AssertMaxInvoke(200, 10, 100, 200);
+            AssertMaxInvoke(0, 0, 0, 0);
+        }
+
+        [MonitoredTest("Should have 3 TextBoxes for entering sides"), Order(4)]
+        public void _4_ShouldHaveThreeTextBoxesForSideInput()
+        {
+            AssertHasTextBoxes();
+        }
+
+        [MonitoredTest("Sides that do not form a triangle should generate an error"), Order(5)]
+        public void _5_SidesThatDoNotFormTriangleShouldGenerateErrorMessages()
+        {
+            AssertHasTextBoxes();
+            Assert.That(_calculateButton, Is.Not.Null); // is provided in template, should be ok
+
+            AssertErrorMessageWithIllegalSides("10", "3", "4");
+            AssertErrorMessageWithIllegalSides("3", "4", "10");
+            AssertErrorMessageWithIllegalSides("3", "10", "4");
+        }
+
+        [MonitoredTest("Sides that form a triangle should produce a correct area"), Order(6)]
+        public void _6_SidesThatFormTriangleShouldProduceCorrectArea()
+        {
+            AssertHasTextBoxes();
+            Assert.That(_calculateButton, Is.Not.Null); // is provided in template, should be ok
+
+            AssertAreaWithLegalSides(6, "3", "4", "5");
+            AssertAreaWithLegalSides(44.039, "9", "10", "12");
+        }
+
+        [MonitoredTest("Mixing illegal sides and legal sides should clear TextBlocks correctly"), Order(7)]
+        public void _7_MixingIllegalSidesWithLegalSidesShouldClearTextBlocksCorrectly()
+        {
+            AssertHasTextBoxes();
+            Assert.That(_calculateButton, Is.Not.Null); // is provided in template, should be ok
+
+            AssertAreaWithLegalSides(6, "3", "4", "5");
+            AssertErrorMessageWithIllegalSides("3", "10", "4");
+            AssertAreaWithLegalSides(44.039, "9", "10", "12");
+        }
+
+        private void AssertAreaWithLegalSides(double expected, string sideA, string sideB, string sideC)
+        {
+            _sideTextBoxes[0].Text = sideA;
+            _sideTextBoxes[1].Text = sideB;
+            _sideTextBoxes[2].Text = sideC;
+
+            _calculateButton.FireClickEvent();
+
+            Assert.That(_errorTextBlock.Text, Is.EqualTo(string.Empty),
+                () => "Sides that form a triangle should not give an error message");
```

Comment:
I think it could be usefull for the students to see the tested sides in the error message (E.g. "Sides that form a triangle (3,5,7) should not give an error message")